// react-fela/lib/ThemeProvider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _felaBindings = require('fela-bindings');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

exports.default = (0, _felaBindings.ThemeProviderFactory)(_react.Component, function (children) {
  return _react.Children.only(children);
}, {
  propTypes: {
    theme: _propTypes2.default.object.isRequired,
    overwrite: _propTypes2.default.bool
  },
  childContextTypes: _defineProperty({}, _felaBindings.THEME_CHANNEL, _propTypes2.default.object),
  contextTypes: _defineProperty({}, _felaBindings.THEME_CHANNEL, _propTypes2.default.object),
  defaultProps: {
    overwrite: false
  }
});
